Firefox/sync-data.js

```diff
@@ -1,16 +1,44 @@
-/* globals require, exports */
+/* globals require, exports, Promise */
 let prefs = require('sdk/preferences/service');
+let {setTimeout} = require('sdk/timers');
 let prefRoot = 'extensions.xkit7.';
 
+function migratePreference(port, prefKey) {
+	let name = prefKey.substring(prefRoot.length);
+	let value = prefs.get(prefKey, null);
+	let msg = {};
+	msg[name] = value;
+	// postMessage doesn't return a promise even though it could :(
+	port.postMessage(msg);
+	return new Promise(function(resolve, reject) {
+		setTimeout(function() {
+			if (port.error) {
+				reject(port.error);
+			} else {
+				resolve();
+			}
+		}, 10);
+	});
+}
+
+function migratePreferences(port, existingKeys) {
+	if (existingKeys.length === 0) {
+		return Promise.resolve();
+	}
+	let key = existingKeys.pop();
+	return migratePreference(port, key).then(function() {
+		return migratePreferences(port, existingKeys);
```

Comment:
mmmm. This seems.... less then optimal. Why use recursion and incur the stack frame penalty here instead of just plain iteration?